### PR TITLE
Convert DynamicStressDivergenceTensors to AD

### PR DESCRIPTION
**Pull request by sveerara**
- Converts DynamicStressDivergenceTensors to AD. 
- Converting InertialForce to AD requires some more work because functions such as adUDotDot, adDuDotDu etc have to be added and I am not sure if any changes are required in the time integrators. So InertialForce is not converted in this PR.
- ADComputeFiniteStrainElasticStress serves as the parent class for one of the materials in MASTODON, so a line is added to the end of the ADComputeFiniteStrainElasticStress to enable this. 

refs #12650

### CODE REVIEW

**Comment by sveerara** [comment]
@dschwen, can you please review this PR? Thanks!

File: modules/tensor_mechanics/include/kernels/ADDynamicStressDivergenceTensors.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  ADDynamicStressDivergenceTensors(const InputParameters & parameters);
+
+protected:
+  ADResidual computeQpResidual();
+
+  const MaterialProperty<RankTwoTensor> & _stress_older;
+  const MaterialProperty<RankTwoTensor> & _stress_old;
+
+  // Rayleigh damping parameter _zeta and HHT time integration parameter _alpha
+  const MaterialProperty<Real> & _zeta;
```
Comment (dschwen) [inline modules/tensor_mechanics/include/kernels/ADDynamicStressDivergenceTensors.h]: If `_zeta` is expected to depend on non-linear variables this line would have to be 

```
const ADMaterialProperty(Real) & _zeta;
```

File: modules/tensor_mechanics/src/materials/ADComputeFiniteStrainElasticStress.C [hunk 2]
```diff
@@ -60,3 +60,6 @@ ADComputeFiniteStrainElasticStress<compute_stage>::computeQpStress()
   // Assign value for elastic strain, which is equal to the mechanical strain
   _elastic_strain[_qp] = _mechanical_strain[_qp];
 }
+
+// explicit instantiation is required for AD base classes
+adBaseClass(ADComputeFiniteStrainElasticStress);
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ADComputeFiniteStrainElasticStress.C]: This should not be necessary, as `ADComputeFiniteStrainElasticStress` gets registered and thus instantiated automatically.

File: modules/tensor_mechanics/test/tests/dynamics/wave_1D/tests [hunk 3]
```diff
@@ -34,6 +34,19 @@
                    " in a linear elastic material with both Rayleigh damping and"
                    " numerical damping resulting from Hilber-Hughes-Taylor (HHT) time integration."
   [../]
+  [./rayleigh_hht_ad]
+    type = 'Exodiff'
```
Comment (lindsayad) [inline modules/tensor_mechanics/test/tests/dynamics/wave_1D/tests]: When doing automatic differentiation work. I generally require that a `PetscJacobianTester` test be added

**Review by lindsayad — DISMISSED** [review]

File: modules/tensor_mechanics/src/materials/ADComputeFiniteStrainElasticStress.C (hunk 2 above)
Comment (sveerara) [reply modules/tensor_mechanics/src/materials/ADComputeFiniteStrainElasticStress.C]: @dschwen, when I comment out the above line and compile MASTODON, i get the following error:`dyld: Symbol not found: __ZN34ADComputeFiniteStrainElasticStressIL12ComputeStage0EE12initialSetupEv`  in all MASTODON tests. I have tried makeclean in both MOOSE and MASTODON and that doesn't seem to help. The MASTODON material model is in https://github.com/sveerara/mastodon/tree/ad_isoil

Any suggestions on what is causing this error?

**Comment by dschwen** [comment]
This is very confusing.

**Comment by lindsayad** [comment]
> This is very confusing.

For implicit instantiations the compiler will only generate symbols for the methods that are used. In registration, that `initialSetup` method is not instantiated so no symbols will be generated for it. However, the explicit instantiation of the class template will generate symbols for all definitions visible in the translation unit ***ahead of the explicit instantiation*** (note that at least on my Mac, `g++` just requires that it be visible in the translation unit, but `clang++` needs the definition ahead of the explicit instantiation). So we  need those lines at the bottom.

File: modules/tensor_mechanics/include/kernels/ADDynamicStressDivergenceTensors.h (hunk 1 above)
Comment (sveerara) [reply modules/tensor_mechanics/include/kernels/ADDynamicStressDivergenceTensors.h]: Thanks for comment, @dschwen. _zeta is an input parameter that can be provided by the user as a function of space but it does not depend on the non-linear variables. So I have not made this change.

File: modules/tensor_mechanics/test/tests/dynamics/wave_1D/tests (hunk 3 above)
Comment (sveerara) [reply modules/tensor_mechanics/test/tests/dynamics/wave_1D/tests]: I have added a PetscJacobianTester test.

**Comment by sveerara** [comment]
@lindsayad and @dschwen, thanks for looking into this issue. 

I did not remove the line at the end of ADComputeFiniteStrainElasticStress

**Comment by permcody** [comment]
@sveerara - We forced in a few commits that failed clang format this weekend. I'm not sure if that's related to the current pre-check failure you are seeing or not, but I bet it is. Can you pull current "devel" and rebase this PR? That should get you past this odd problem.

**Comment by sveerara** [comment]
@permcody, that fixed the clang-format test. Thanks!

File: modules/tensor_mechanics/doc/content/source/kernels/ADDynamicStressDivergenceTensors.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# ADDynamicStressDivergenceTensors
+
+!syntax description /Kernels/ADDynamicStressDivergenceTensors
+
+## Description
+
+This class computes the stress divergence and the stiffness proportional Rayleigh damping for structural dynamics problems. Each ADDynamicStressDivergenceTensors input block computes force in one direction. So, a separate ADDynamicStressDivergenceTensors input block should be set up for each coordinate direction. The [DynamicTensorMechanics](/DynamicTensorMechanicsAction.md) action automatically sets up the ADDynamicStressDivergenceTensors input block in all coordinate direction. More information about the residual calculation and usage can be found at [Dynamics](Dynamics.md).
```
Comment (dschwen) [inline modules/tensor_mechanics/doc/content/source/kernels/ADDynamicStressDivergenceTensors.md]:
```suggestion
This class computes the stress divergence and the stiffness proportional Rayleigh damping for structural dynamics problems. Each ADDynamicStressDivergenceTensors input block computes force in one direction. So, a separate ADDynamicStressDivergenceTensors input block should be set up for each coordinate direction. The [DynamicTensorMechanics](/DynamicTensorMechanicsAction.md) action automatically sets up the ADDynamicStressDivergenceTensors input block in all coordinate directions. More information about the residual calculation and usage can be found at [Dynamics](Dynamics.md).
```

Comment (dschwen) [reply modules/tensor_mechanics/doc/content/source/kernels/ADDynamicStressDivergenceTensors.md]: P.S.: This is one reason I prefer hard wrapped .md files (and .tex files). Spot the change!

File: modules/tensor_mechanics/src/kernels/ADDynamicStressDivergenceTensors.C [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+#include "ADDynamicStressDivergenceTensors.h"
+#include "ElasticityTensorTools.h"
+
+registerADMooseObject("TensorMechanicsApp", ADDynamicStressDivergenceTensors);
+
+defineADValidParams(
+    ADDynamicStressDivergenceTensors,
+    ADStressDivergenceTensors,
+    params.addClassDescription(
+        "Residual due to stress related Rayleigh damping and HHT time integration terms ");
```
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/ADDynamicStressDivergenceTensors.C]:
```suggestion
        "Residual due to stress related Rayleigh damping and HHT time integration terms");
```

File: modules/tensor_mechanics/src/kernels/ADDynamicStressDivergenceTensors.C [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+   * The residual of _zeta*K*[(1+_alpha)vel-_alpha vel_old]+ alpha K [ u - uold] + K u is required
+   * = _zeta*[(1+_alpha)d/dt (Div sigma)-alpha d/dt(Div sigma_old)] +alpha [Div sigma - Div
+   *sigma_old]+ Div sigma
+   * = _zeta*[(1+alpha)(Div sigma - Div sigma_old)/dt - alpha (Div sigma_old - Div sigma_older)/dt]
+   *   + alpha [Div sigma - Div sigma_old] +Div sigma
+   * = [(1+_alpha)*_zeta/dt +_alpha+1]* Div sigma - [(1+2_alpha)*_zeta/dt + _alpha] Div sigma_old +
+   *_alpha*_zeta/dt Div sigma_older
+   */
+
+  ADResidual residual = 0.0;
```
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/ADDynamicStressDivergenceTensors.C]:
```suggestion
  ADResidual residual;
```

File: modules/tensor_mechanics/src/kernels/ADDynamicStressDivergenceTensors.C [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  if (_static_initialization && _t == _dt)
+  {
+    // If static inialization is true, then in the first step residual is only Ku which is
+    // stress.grad(test).
+    residual += _stress[_qp].row(_component) * _grad_test[_i][_qp];
```
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/ADDynamicStressDivergenceTensors.C]:
```suggestion
    residual = _stress[_qp].row(_component) * _grad_test[_i][_qp];
```

File: modules/tensor_mechanics/src/kernels/ADDynamicStressDivergenceTensors.C [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+    if (_volumetric_locking_correction)
+      residual +=
+          _stress[_qp].trace() / 3.0 * (_avg_grad_test[_i] - _grad_test[_i][_qp](_component));
+  }
+  else if (_dt > 0)
+  {
+    residual +=
```
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/ADDynamicStressDivergenceTensors.C]:
```suggestion
    residual =
```

Comment (dschwen) [reply modules/tensor_mechanics/src/kernels/ADDynamicStressDivergenceTensors.C]: This looks petty, but an assignment might actually be faster than an unnecessary addition to a zero dual number.

File: modules/tensor_mechanics/test/tests/dynamics/wave_1D/tests [hunk 9]
```diff
@@ -34,6 +34,31 @@
...
+    input = 'wave_rayleigh_hht_AD.i'
+    exodiff = 'wave_rayleigh_hht_out.e'
+    abs_zero = 1e-09
+    compiler = 'GCC CLANG'
```
Comment (dschwen) [inline modules/tensor_mechanics/test/tests/dynamics/wave_1D/tests]: do we still need this restriction? We dropped intel compiler support a few months ago.

File: modules/tensor_mechanics/test/tests/dynamics/wave_1D/tests [hunk 10]
```diff
@@ -34,6 +34,31 @@
...
+    type = 'PetscJacobianTester'
+    input = 'wave_rayleigh_hht_AD.i'
+    compiler = 'GCC CLANG'
+
+    requirement = "The mechanics system shall correctly compute the Jacobian"
+                 " for 1D wave propagation in a linear elastic material with"
+                 " both Rayleigh damping and numerical damping resulting from"
+                 " Hilber-Hughes-Taylor (HHT) time integration when automatic"
+                 " differentiation is used."
+    prereq = "rayleigh_hht_ad"
```
Comment (dschwen) [inline modules/tensor_mechanics/test/tests/dynamics/wave_1D/tests]: The pattern in many tests now is to suppress outputs in the jacobian test. Then you should be able to run it concurrently with the exidiff test.

**Review by dschwen — CHANGES_REQUESTED** [review]
Your `InertialForce` kernel depends on velocity and acceleration, which are AuxVariables that have no derivatives (and probably shouldn't because they are previous' time step values). Is this a lagging scheme? (just my curiosity)

File: modules/tensor_mechanics/doc/content/source/kernels/ADDynamicStressDivergenceTensors.md (hunk 4 above)
Comment (lindsayad) [reply modules/tensor_mechanics/doc/content/source/kernels/ADDynamicStressDivergenceTensors.md]: What do you mean by hard wrapped? An enforced line length?

Comment (dschwen) [reply modules/tensor_mechanics/doc/content/source/kernels/ADDynamicStressDivergenceTensors.md]: Sorry that was a typo, I meant "hard rapped".

File: modules/tensor_mechanics/test/tests/dynamics/wave_1D/tests (hunk 10 above)
Comment (lindsayad) [reply modules/tensor_mechanics/test/tests/dynamics/wave_1D/tests]: We actually explicit set `Outputs/exodus=false` by default with the `PetscJacobianTester` now. You can set `turn_off_exodus_output=True` in the test spec if you do want the exodus output. This way of turning off exodus output from the command line would not save you if you have an exodus sub-block in the Outputs block, however.

File: modules/tensor_mechanics/doc/content/source/kernels/ADDynamicStressDivergenceTensors.md (hunk 4 above)
Comment (dschwen) [reply modules/tensor_mechanics/doc/content/source/kernels/ADDynamicStressDivergenceTensors.md]: Yeah, of course, hard line breaks before the line length exceeds 80 chars. Most editors have a reflow hotkey, don't they?

Comment (lindsayad) [reply modules/tensor_mechanics/doc/content/source/kernels/ADDynamicStressDivergenceTensors.md]: > Sorry that was a typo, I meant "hard rapped".

You like hard rap? 😄 

> line length exceeds 80 chars

80? Is that a standard of some kind? I personally use 100 for consistency with our clang-format, and yes I set that with my editor.

Comment (dschwen) [reply modules/tensor_mechanics/doc/content/source/kernels/ADDynamicStressDivergenceTensors.md]: https://hackernoon.com/does-column-width-of-80-make-sense-in-2018-50c161fbdcf6

80 chars is an ancient convention! And you can debate (see above) whether it makes sense for code with wider screens. But I think the wider screens argument is misguided. Just use multiple editor views side by side. Text with longer lines is simply less readable, regardless of screen width:

https://practicaltypography.com/line-length.html

**Comment by lindsayad** [comment]
git blame aside would you argue for a change in our .clang-format file? I’m guessing there was a debate about this at the inception of clang-format.

**Comment by dschwen** [comment]
Hm, I'm not advocating to change our code style standard, I'd just like to have some formatting convention for md files. Those line lengths don't necessarily have to be the same.

**Review by dschwen — DISMISSED** [review]
👍

**Comment by sveerara** [comment]
@dschwen, thanks for your comments on the PR. I have addressed your comments. 

Regarding your comment on InertialForce kernel, it currently works both using velocity/acceleration auxvariables and also using the derivatives computed by the timeintegrator (if the auxvariables are not provided as input). The aim is to deprecate the aux variables version once all the necessary support objects (such as postprocessors to print out first or second time derivatives) are available for the time integrator version.

**Comment by dschwen** [comment]
Uh oh, the failure is my fault! I didn't pay close enough attention and missed that there is a code path that leaves `residual` uninitialized! Sorry, but add an `else residual = 0.0;` before the return.

**Comment by sveerara** [comment]
@dschwen, thanks for catching the error. I should have compiled it on my laptop before updating the PR. All tests are passing now.

**Review by dschwen — APPROVED** [review]

**Pull request merged**